State Channel example contract (#3)
keypair/white-paper: sophia.tex

arrives is outside the scope of the blockchain

\subsubsection{State Channel example}

Contracts can be deployed on-chain or in a state channel. State channels
provide means for two parties for executing transactions off-chain. Once the
channel is opened on-chain all the communication is moved off-chain and it is
just between the two participants. It is not limited by blockchain
throughputh, fees or gas costs. This provides a new environment to run
contracts in. In order to make most out of it, off-chain contracts can use
on-chain data.

An example of the use of off-chain contract and an on-chain oracle can be

> illustrated with a little insurance example. Note that all computation on a

Reviewer: ```suggestion
illustrated with a simple insurance contract. Note that all computation on a
```

blockchain must be deterministic in order to be able to validate the results.
If not exactly the same, the corresponding state hashes will differ. This is
true especially with state channels as contracts are executed by both
participants in their own context. This requires both to share the same view
of the chain\footnote{We achieve this by specifying the on-chain environment
as part of the off-chain update.}.

Running a simple insurance contract in which one user provides the service
while the other insures their property for a certain timeframe from hailstorm
requires some input from real life data. This is where oracle input comes: for

> the example we have an on-chain Oracle to produce data if there had been a

Reviewer: ```suggestion
the example we have an on-chain Oracle to produce data on whether there has been a
```

hailstorm in a certain city. It publishes data pinned in time with a

> corresponding block height when it occurred. Later on insured participant can

Reviewer: ```suggestion
corresponding block height when it occurred. Later on insured participants can
```

Author: Well one is insuring the other, so only one is being insured 🤔

Reviewer: Ok, then it should be "the insured participant"

Author: Addressed

    if (Chain.block_height =< state.active_to) // insured
      require(Contract.balance - amount >= state.compensation, "not_enough_compensation")
    Chain.spend(Contract.creator, amount)

  payable entrypoint deposit() =
    require(Call.caller == Contract.creator, "not_service_provider")

  entrypoint get_insurance_range() =
    (state.active_from, state.active_to)

  stateful entrypoint claim_insurance(q: query_id) =
    require(Call.caller != Contract.creator, "service_provider")
    switch(Oracle.get_answer(state.oracle, q))
      None         =>
        abort("no_response")
      Some(was_there_hailstorm) =>
        let (city, generation) = Oracle.get_question(state.oracle, q)
        require(city == state.city, "different_city")
        let is_in_range = generation >= state.active_from && generation =< state.active_to
        require(is_in_range, "not_in_insurance_range")
        require(was_there_hailstorm, "different_response")
        Chain.spend(Call.caller, state.compensation)
        put(state{ active_from = 0,
                   active_to = 0})

\end{verbatim}

The contract is deployed by the insurance service provider in a state channel.

> It stores the trusted oracle in its state, the price it expects for insurance

Reviewer: stores a reference to?

for a single generation and in which city the insured city belongs to. It also
contains the insured interval in block heights, which initially is set to 0.

The creator of this contract may now deposit initial amount for compensation.

\begin{verbatim}
  payable entrypoint deposit() =
    require(Call.caller == Contract.creator, "not_service_provider")
\end{verbatim}

Note that there is a check there: only the owner of the contract - being the
service provider - can execute that function. Since the insured participant is
not expected to call that function - any attempt of doing so will result in
aborting the call. This is true both off-chain and on-chain in a forced
progress scenario.

The keyword \verb+payable+ expresses that we expect the caller to add a token
amount to the contract call transaction, which could be checked by comparing
\verb+Call.value+. There is no restriction for the amount being deposited in
the contract and this could be done in a couple of different steps if needed.

> Once that the compensation is present in the contract, the other party can

Reviewer: ```suggestion
Once the compensation is present in the contract, the other party can
```

call the \verb+insure+ function.

\begin{verbatim}
  stateful payable entrypoint insure() =
    require(Call.caller != Contract.creator, "service_provider")
    require(Contract.balance + Call.value >= state.compensation, "not_enough_compensation")
    let period : int = Call.value / state.price_per_gen
    if (Chain.block_height >= state.active_to) // expired, renew
      put(state{ active_from = Chain.block_height,
                 active_to = Chain.block_height + period})
    else // not expired, extend
      put(state{ active_to = state.active_to + period })
\end{verbatim}

There are some checks that it is not the insurance provider that is calling
the contract and that the contract has enough tokens to cover the compensation
in case of a hailstorm. If those pass the corresponding insurance period is
calculated according to the tokens dedicated in the contract and the state is
updated accordingly. Note that if there is an active insurance at the moment,

> existing time frame is just extended. This allows for continuous insurances.

Reviewer: ```suggestion
the existing time frame is just extended. This allows for continuous insurances.
```

Once an insurance event takes place, the insured party simply claims their
due, providing the Oracle's response.

\begin{verbatim}
  stateful entrypoint claim_insurance(q: query_id) =
    require(Call.caller != Contract.creator, "service_provider")
    switch(Oracle.get_answer(state.oracle, q))
      None         =>
        abort("no_response")
      Some(was_there_hailstorm) =>
        let (city, generation) = Oracle.get_question(state.oracle, q)
        require(city == state.city, "different_city")
        let is_in_range = generation >= state.active_from && generation =< state.active_to
        require(is_in_range, "not_in_insurance_range")
        require(was_there_hailstorm, "different_response")
        Chain.spend(Call.caller, state.compensation)
        put(state{ active_from = 0,
                   active_to = 0})
\end{verbatim}

> Note that if any of the check fails, the whole execution of the contract

Reviewer: ```suggestion
Note that if any of the checks fail, the whole execution of the contract
```
